velox/common/memory/Memory.cpp

```diff
@@ -60,13 +68,51 @@ uint16_t MemoryManager::alignment() const {
   return alignment_;
 }
 
-MemoryPool& MemoryManager::getRoot() const {
-  return *root_;
+std::shared_ptr<MemoryPool> MemoryManager::getPool(
+    const std::string& name,
+    MemoryPool::Kind kind,
+    int64_t cap) {
+  std::string poolName = name;
+  if (poolName.empty()) {
+    static std::atomic<int64_t> poolId{0};
+    poolName =
+        fmt::format("default_{}_{}", MemoryPool::kindString(kind), poolId++);
+  }
+  if (kind == MemoryPool::Kind::kLeaf) {
+    return defaultRoot_->addChild(poolName, kind);
+  }
+
+  MemoryPool::Options options;
+  options.alignment = alignment_;
+  options.capacity = cap;
+  auto pool = std::make_shared<MemoryPoolImpl>(
+      this,
+      poolName,
+      MemoryPool::Kind::kAggregate,
+      nullptr,
+      poolDestructionCb_,
+      options);
+  folly::SharedMutex::WriteHolder guard{mutex_};
```

Comment:
Why not `std::lock_guard<std::mutex>` ?

Reply:
Use r/w lock here for memory arbitration support. Each memory arbitration request holds a reader lock to allow concurrent access to the memory arbitration (the memory arbitrartor can serialize the arbitration requests internally) and the thread safety of the memory pools_. Also user can access memory manager like check the number of pools without blocking.